### PR TITLE
[BOP-409] [BOP-1114] Cloud Providers

### PR DESCRIPTION
**Pull request by tppolkow**

### CODE REVIEW

**Review by KoryKessel-Mirantis — CHANGES_REQUESTED** [review]

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+---
+title: CloudProvider
+weight: 6
+---
+
+Enabling CloudProvider allows your MKE cluster to integrate with cloud provider addons. Currently, only AWS is supported as an MKE managed cloud provider addon.
+To use other cloud providers, you must set MKE to run with "external" cloud provider and install the cloud provider yourself.
```
[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md]: done

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Pre-requisites
+Cloud Providers may have additional requirements. Please refer to the specific cloud provider documentation for more information.
```
[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md]: done

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+---
+title: CloudProvider
+weight: 6
+---
+
+Enabling CloudProvider allows your MKE cluster to integrate with cloud provider addons. Currently, only AWS is supported as an MKE managed cloud provider addon.
```
[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md]: @KoryKessel-Mirantis The first sentence you added doesn't make sense here, I believe you are defining the low level[ Kubernetes cloud provider](https://github.com/kubernetes/cloud-provider), but this is an interface designed to be used by developers developing their own cloud provider addons ( hence the `shared interfaces that are implemented by Kubernetes cloud service providers` ). It doesn't really make sense here as here they are enabling the "implemented interface" i.e the AWS cloud provider (which implements the generic Kubernetes defined interface)

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+To use other cloud providers, you must set MKE to run with "external" cloud provider and install the cloud provider yourself.
+
+## Pre-requisites
```
[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md]: done

File: content/docs/operations/cloudprovider/cloudprovider.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md]: Workshopped the paragraph to goodness in Slack and tapped the agreed-to edit into my suggested change.

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+{{< /callout >}}
+
+## Prerequisites
+Refer to the documentation for your chosen cloud service provider to learn of any proprietary requirements.
+
+For using the MKE managed AWS Cloud Provider you will need to ensure your nodes have certain IAM policies. See the official AWS cloud provider [documentation](https://cloud-provider-aws.sigs.k8s.io/prerequisites/#iam-policies).
+
+## Configuration
+
+CloudProvider support is disabled by default. To enable CloudProvider support, configure
```
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md:19]: Still pending?

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+```yaml
+  cloudProvider:
+    enabled: true
+    provider: aws
+```
+
+**Configuration parameters**
```
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md:27]: Still pending?

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
+---
+title: CloudProvider
```
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md:2]: Still pending?

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+With MKE 4, you can deploy a cloud provider to integrate your MKE cluster with cloud provider service APIs.
+
+{{< callout type="note" >}}
+AWS is currently the only managed cloud service provider add-on supported by MKE 4. To use a different cloud service provider, you must set MKE 4 to run with "external" cloud provider and install the desired cloud provider.
+{{< /callout >}}
+
+## Prerequisites
+Refer to the documentation for your chosen cloud service provider to learn of any proprietary requirements.
+
+For using the MKE managed AWS Cloud Provider you will need to ensure your nodes have certain IAM policies. See the official AWS cloud provider [documentation](https://cloud-provider-aws.sigs.k8s.io/prerequisites/#iam-policies).
```
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md:15]: Still pending?

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+**Configuration parameters**
+
+| Field    | Description                                                                                                             | Default |
+|----------|-------------------------------------------------------------------------------------------------------------------------|---------|
+| enabled  | Enables cloud provider flags on mke components.                                                                         | false   |
+| provider | Either "aws" or "external". If "external" is specified the user is responsible for installing their own cloud provider. | ""      |
+
+
+## Example: Create an NLB with Cloud Provider AWS
```
[the remark this answers is not in the document]
Comment (KoryKessel-Mirantis) [reply content/docs/operations/cloudprovider/cloudprovider.md:35]: Still pending?

**Review by KoryKessel-Mirantis — CHANGES_REQUESTED** [review]

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
+---
+title: CloudProvider
+weight: 6
+---
+
+With MKE 4, you can deploy a cloud provider to integrate your MKE cluster with cloud provider service APIs.
+
+{{< callout type="note" >}}
+AWS is currently the only managed cloud service provider add-on supported by MKE 4. To use a different cloud service provider, you must set MKE 4 to run with "external" cloud provider and install the desired cloud provider.
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:9]:
```suggestion
AWS is currently the only managed cloud service provider add-on supported by MKE 4. To use a different cloud service provider, before installing that provider you must change the `provider` parameter under `cloudProvider` in the MKE configuration file to `external`:

```yaml
  cloudProvider:
    enabled: true
    provider: aws
```
```

change the `enabled` parameter under `cloudProvider` in the MKE configuration file to `true`:

**Review by KoryKessel-Mirantis — CHANGES_REQUESTED** [review]

File: content/docs/operations/cloudprovider/cloudprovider.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md:40]: @KoryKessel-Mirantis Cloud Provider has many uses that are up to the user. This example is just one way users could use cloud provider. I feel your edit could make users think this is an integral part of cloud provider set up, which it is not.

[the remark this answers is not in the document]
Comment (tppolkow) [reply content/docs/operations/cloudprovider/cloudprovider.md]: commiting this given the first line of the section says its an example.

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx-container
+        image: nginx:latest
+        ports:
+        - containerPort: 80
+EOF
+```
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:70]:
```suggestion
1. Create a sample nginx deployment:

   ```shell
   cat <<EOF | kubectl --kubeconfig ~/.mke/mke.kubeconf apply -f -
   apiVersion: apps/v1
   kind: Deployment
   metadata:
     name: nginx-deployment
   spec:
     replicas: 3  
     selector:
       matchLabels:
         app: nginx
     template:
       metadata:
         labels:
           app: nginx
       spec:
         containers:
         - name: nginx-container
           image: nginx:latest
           ports:
           - containerPort: 80
   EOF
   ```
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+spec:
+  selector:
+    app: nginx
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 80
+  type: LoadBalancer
+EOF
+```
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:90]:
```suggestion
2. Create a service of type LoadBalancer:

   ```shell
   cat <<EOF | kubectl --kubeconfig ~/.mke/mke.kubeconf apply -f -
   apiVersion: v1
   kind: Service
   metadata:
     name: nginx-service
     annotations:
       service.beta.kubernetes.io/aws-load-balancer-type: nlb
   spec:
     selector:
       app: nginx
     ports:
       - protocol: TCP
         port: 80
         targetPort: 80
     type: LoadBalancer
   EOF
   ```
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+3. Check the status of the service
+```shell
+kubectl --kubeconfig ~/.mke/mke.kubeconf get service
+NAME            TYPE           CLUSTER-IP     EXTERNAL-IP                                                                        PORT(S)        AGE
+kubernetes      ClusterIP      10.96.0.1      <none>                                                                             443/TCP        14m
+nginx-service   LoadBalancer   10.96.177.89   afdf81e0681274c52acbb7b45add87a1-637d0d850105ea92.elb.ca-central-1.amazonaws.com   80:32927/TCP   63s
+```
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:98]:
```suggestion
3. Check the status of the service:

   ```shell
   kubectl --kubeconfig ~/.mke/mke.kubeconf get service
   NAME            TYPE           CLUSTER-IP     EXTERNAL-IP                                                                        PORT(S)        AGE
   kubernetes      ClusterIP      10.96.0.1      <none>                                                                             443/TCP        14m
   nginx-service   LoadBalancer   10.96.177.89   afdf81e0681274c52acbb7b45add87a1-637d0d850105ea92.elb.ca-central-1.amazonaws.com   80:32927/TCP   63s
   ```
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+4. Load balancer should be created and visible in AWS console.
+![aws-lb.png](aws-lb.png)
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:101]:
```suggestion
The load balancer should now be visible in AWS console.

![aws-lb.png](aws-lb.png)
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+   
+
+5. After some time, the load balancer should finish provisioning and we should be able to access nginx via the external IP.
+![aws-lb-provisioned.png](aws-lb-provisioned.png)
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:105]:
```suggestion
Once the load balancer finishes provisioning you should be able to access nginx through the external IP.

![aws-lb-provisioned.png](aws-lb-provisioned.png)
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+|----------|-------------------------------------------------------------------------------------------------------------------------|---------|
+| enabled  | Enables cloud provider flags on mke components.                                                                         | false   |
+| provider | Either "aws" or "external". If "external" is specified the user is responsible for installing their own cloud provider. | ""      |
+
+
+## Example: Create an NLB with Cloud Provider AWS
+
+Here is an example of how to use cloud provider AWS to create a Network Load Balancer (NLB) in your MKE cluster. 
+
+Once you have enabled the cloud provider in your MKE configuration, and applied the mke configuration, you can create an NLB using the following steps:
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:44]:
```suggestion
An example is offered below of how to use cloud provider AWS to create a Network Load Balancer (NLB) in your MKE cluster. 

Once you have enabled the cloud provider through the MKE configuration file and have applied it, you can create an NLB as follows:
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+AWS is currently the only managed cloud service provider add-on supported by MKE 4. To use a different cloud service provider, before installing that provider you must change the `provider` parameter under `cloudProvider` in the MKE configuration file to `external`:
+
+```yaml
+  cloudProvider:
+    enabled: true
+    provider: aws
+{{< /callout >}}
+
+## Prerequisites
+Refer to the documentation for your chosen cloud service provider to learn of any proprietary requirements.
```
Comment (KoryKessel-Mirantis) [inline content/docs/operations/cloudprovider/cloudprovider.md:18]:
```suggestion
## Prerequisites

Refer to the documentation for your chosen cloud service provider to learn of any proprietary requirements.
```

**Review by KoryKessel-Mirantis — CHANGES_REQUESTED** [review]

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+   NAME            TYPE           CLUSTER-IP     EXTERNAL-IP                                                                        PORT(S)        AGE
+   kubernetes      ClusterIP      10.96.0.1      <none>                                                                             443/TCP        14m
+   nginx-service   LoadBalancer   10.96.177.89   afdf81e0681274c52acbb7b45add87a1-637d0d850105ea92.elb.ca-central-1.amazonaws.com   80:32927/TCP   63s
+   ```
+
+The load balancer should now be visible in AWS console.
+
+![aws-lb.png](aws-lb.png)
+
+Once the load balancer finishes provisioning you should be able to access nginx through the external IP.
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:109]:
```suggestion
Once the load balancer finishes provisioning, you should be able to access nginx through the external IP.
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 19]
```diff
@@ -0,0 +1,111 @@
+---
+title: Cloud providers
+weight: 6
+---
+
+With MKE 4, you can deploy a cloud provider to integrate your MKE cluster with cloud provider service APIs.
+
+{{< callout type="note" >}}
+AWS is currently the only managed cloud service provider add-on supported by MKE 4. To use a different cloud service provider, before installing that provider you must change the `provider` parameter under `cloudProvider` in the MKE configuration file to `external`:
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:9]:
```suggestion
To install a different cloud service provider, you must first change the `provider` parameter
under `cloudProvider` in the MKE configuration file to `external` before proceeding to the installation:
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+```yaml
+  cloudProvider:
+    enabled: true
+    provider: aws
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:14]:
```suggestion
    provider: external
```
To match description above

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+```
+{{< /callout >}}
+
+## Prerequisites
+
+Refer to the documentation for your chosen cloud service provider to learn of any proprietary requirements.
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:20]:
```suggestion
Refer to the documentation of your chosen cloud service provider to learn about any proprietary requirements.
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+
+| Field      | Description                                                                                                             | Default   |
+|------------|-------------------------------------------------------------------------------------------------------------------------|-----------|
+| `enabled`  | Enables cloud provider flags on MKE components.                                                                         | `false`   |
+| `provider` | Either `aws` or `external`. If "external" is specified the user is responsible for installing their own cloud provider. | ""    ``  |
+
+
+## Create an NLB with AWS Cloud Provider
+
+An example is offered below of how to use cloud provider AWS to create a Network Load Balancer (NLB) in your MKE cluster.
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:44]:
```suggestion
Below example ilustrates how to use cloud provider AWS to create a Network Load Balancer (NLB) in your MKE cluster. 
```

File: content/docs/operations/cloudprovider/cloudprovider.md [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+       spec:
+         containers:
+         - name: nginx-container
+           image: nginx:latest
+           ports:
+           - containerPort: 80
+   EOF
+   ```
+
+2. Create a service of type LoadBalancer:
```
Comment (MagdaDziadosz) [inline content/docs/operations/cloudprovider/cloudprovider.md:75]:
```suggestion
2. Create a service of type `LoadBalancer`:
```

**Review by MagdaDziadosz — CHANGES_REQUESTED** [review]

**Review by MagdaDziadosz — APPROVED** [review]

**Review by KoryKessel-Mirantis — APPROVED** [review]

**Pull request merged**